lrpar/src/lib/lex.rs

```diff
@@ -69,18 +83,23 @@ impl<StorageT: Copy> Lexeme<StorageT> {
         self.start
     }
 
-    /// Byte offset of the start of the lexeme
-    pub fn end(&self) -> usize {
-        self.start() + self.len()
-    }
-
-    /// Length in bytes of the lexeme
-    pub fn len(&self) -> usize {
-        debug_assert!(size_of::<usize>() >= size_of::<u32>());
-        self.len as usize
+    /// Byte offset of the end of the lexeme, or `None` if this lexeme is the result of error
+    /// recovery.
+    pub fn end(&self) -> Option<usize> {
+        if let Some(l) = self.len() {
+            Some(self.start() + l)
+        } else {
+            None
+        }
     }
 
-    pub fn is_empty(&self) -> bool {
-        self.len == 0
+    /// Length in bytes of the lexeme, or `None` if this lexeme is the result of error recovery.
+    pub fn len(&self) -> Option<usize> {
+        if self.len == u32::max_value() {
```

Comment:
Can this restriction have any negative consequences to the user, e.g. when they have a lexeme of exactly u32 length? Should we document this somewhere?

Reply:
Yes, it definitely is restrictive: they can only have lexemes of (4 gigabytes - 1 byte). I must admit that I don't think we'll hit that case for a while and, if/when we do, we can transparently change the storage of `Lexeme` to store more bits. It's one of those "I should fix this, but it's low priority" things.

Reply:
I agree, it's unlikely someone will run into this, but you never know. Isn't the fix just changing `self.len` to `Option` as well? Or am I missing something?

Reply:
Changing it to `Option` will add a whole extra machine word to this struct, which is pretty punishing for the case of generic parse trees. The good news is that, if someone does create an input with a single lexeme of 4GiB, it will at least crash gracefully at the first possible opportunity.

Reply:
Ah! I hadn't considered the memory implication of that. I agree that in this case it's not worth it then.